### ISSUE TITLE
Potential Issue with `scatter_and_pack_data_list` Leading to Incorrect Value Prediction in Critic Model

### ISSUE THREAD

**Issue by valorix25** [ChenmienTan/RL2#16]
### Description:

The method `minibatches = self.scatter_and_pack_data_list(data_list)` groups the `data_list` and concatenates each group into a single long sequence of data. This introduces a potential issue: 

In `minibatches[0]['action_mask'].tolist()`, there are interleaved segments of consecutive `0`s and `1`s. For example, the second segment of consecutive `0`s actually corresponds to the second trajectory in the original data. However, it can still "see" the preceding consecutive `1`s, which are associated with the `input_ids`. 

Could this behavior potentially affect the value prediction when using the `Critic.model`? Specifically, does the interleaving of masked and unmasked data influence the model's ability to correctly predict values for individual trajectories?

Here is my observation:

<img width="835" height="558" alt="Image" src="https://github.com/user-attachments/assets/6e6dcf90-19db-47c7-ae8e-e5fe947d3a64" />

Here is the related code:

```Python
class Critic(Worker):

    def __init__(self, config):
        super().__init__(config, True)

        self.model = AutoModelForTokenClassification.from_pretrained(
            config.model_name,
            num_labels=1,
            trust_remote_code=True,
            attn_implementation="flash_attention_2"
        )

        if hasattr(self.config, "lora") and self.config.lora.rank > 0:
            assert config.tp_size == 1, \
                "LoRA is not compatible with tensor parallelism."
            self.model = prepare_lora_model(
                self.model, "TOKEN_CLS", config.lora
            )

        self.prepare_model_optimizer()

    def forward(self, minibatch) -> torch.Tensor:
        update_params_of_ring_attn(
            minibatch["cu_seqlens"], self.device_mesh["sp"]
        )

        return self.model(
            input_ids=minibatch["states"],
            position_ids=minibatch["position_ids"],
            use_cache=False
        ).logits.squeeze(-1) * minibatch["action_mask"]

    @time_logger("compute_values")
    @torch.no_grad()
    def compute_values(self, data_list, step):
        self.load_model_to_gpu()
        import pdb; pdb.set_trace()  # Debugging breakpoint
        minibatches = self.scatter_and_pack_data_list(data_list)

        self.model.eval()
        for minibatch in self.tqdm(minibatches, desc="Compute values"):
            minibatch["values"] = self.forward(minibatch)
        
        self.offload_model_to_cpu()
        return self.unpack_and_gather_data_list(minibatches)

```

**Comment by ChenmienTan** [comment]
`update_param_of_ring_attn` will mask attention across trajectories

Reference: https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/flash_attn_interface.py#L143
